jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/TransformingFlusher.java

```diff
@@ -0,0 +1,207 @@
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.websocket.core.Frame;
+
+/**
+ * This is used to iteratively transform or process a frame.
+ */
+public abstract class TransformingFlusher
+{
+    private final Logger log = Log.getLogger(this.getClass());
+
+    private final Queue<FrameEntry> entries = new ArrayDeque<>();
+    private final IteratingCallback flusher = new Flusher();
+    private boolean finished = true;
+    private Throwable failure;
+
+    /**
+     * Called when a frame is ready to be transformed.
+     * @param frame the frame to transform.
+     * @param callback used to signal to start processing again.
+     * @param batch whether this frame can be batched.
+     */
+    protected abstract void onFrame(Frame frame, Callback callback, boolean batch);
+
+    /**
+     * Called multiple times to transform the frame given in {@link TransformingFlusher#onFrame(Frame, Callback, boolean)}.
+     * @param callback used to signal to start processing again.
+     */
+    protected abstract void transform(Callback callback);
+
+    /**
+     * Called to indicate that you have finished transforming this frame and are ready to receive a new one
+     * after the next the callback is succeeded.
+     * @param finished
+     */
+    public final void setFinished(boolean finished)
+    {
+        this.finished = finished;
+    }
+
+    public final void sendFrame(Frame frame, Callback callback, boolean batch)
+    {
+        FrameEntry entry = new FrameEntry(frame, callback, batch);
+        if (log.isDebugEnabled())
+            log.debug("Queuing {}", entry);
+
+        boolean enqueued = false;
+        synchronized (this)
+        {
+            if (failure == null)
+            {
+                enqueued = entries.add(entry);
+            }
+        }
+
+        if (enqueued)
+            flusher.iterate();
+        else
+            notifyCallbackFailure(callback, failure);
+    }
+
+    private void onFailure(Throwable t)
+    {
+        synchronized (this)
+        {
+            if (failure == null)
+                failure = t;
+        }
+
+        for (FrameEntry entry : entries)
+            notifyCallbackFailure(entry.callback, t);
+        entries.clear();
+    }
+
+    private FrameEntry pollEntry()
+    {
+        synchronized (this)
+        {
+            return entries.poll();
+        }
+    }
+
+    private class Flusher extends IteratingCallback implements Callback
+    {
+        private FrameEntry current;
+
+        @Override
+        protected Action process()
+        {
+            if (finished)
+            {
+                current = pollEntry();
+                if (log.isDebugEnabled())
+                    log.debug("Polled {}", current);
+                if (current == null)
+                    return Action.IDLE;
+            }
+
+            if (log.isDebugEnabled())
+                log.debug("Processing {}", current);
+
+            if (finished)
```

Comment:
don't repeat `if(finished)` just for debugging, have multiple debugs if needed and a simple logic flow.

Reply:
It's a bit weird that you start with `finished==true` to know when to call `onFrame` or `transform`.  It would be better to call `onFrame` IFF you just did a `pollEntry` for a new value of `current`